diagnostics/diagnostics_manager.rs

```diff
@@ -0,0 +1,217 @@
+        self.error_code = Some(error_code);
+    }
+}
+
+impl<'d> Drop for DiagnosticsResultTracker<'d> {
+    fn drop(&mut self) {
+        match &self.error_code {
+            Some(error_code) => {
+                self.diagnostics_manager.submit_action_fail(self.database_name, self.action_kind);
+                self.diagnostics_manager.submit_error(self.database_name, error_code.clone());
+            }
+            None => self.diagnostics_manager.submit_action_success(self.database_name, self.action_kind),
+        }
+    }
+}
+
+pub fn run_with_diagnostics<F, T>(
+    diagnostics_manager: &DiagnosticsManager,
+    action_kind: ActionKind,
+    f: F,
+) -> Result<T, Status>
+where
+    F: FnOnce() -> Result<T, Status>,
+{
+    run_and_track(DiagnosticsResultTracker::without_database(diagnostics_manager, action_kind), f)
+}
+
+pub fn run_with_diagnostics_for_database<F, T>(
+    diagnostics_manager: &DiagnosticsManager,
+    database_name: &str,
+    action_kind: ActionKind,
+    f: F,
+) -> Result<T, Status>
+where
+    F: FnOnce() -> Result<T, Status>,
+{
+    run_and_track(DiagnosticsResultTracker::with_database(diagnostics_manager, database_name, action_kind), f)
+}
+
+fn run_and_track<F, T>(mut tracker: DiagnosticsResultTracker<'_>, f: F) -> Result<T, Status>
+where
+    F: FnOnce() -> Result<T, Status>,
+{
+    let result = f();
+    if let Err(ref status) = result {
+        tracker.set_error_code(get_status_error_code(status));
+    }
+    result
+}
+
+pub async fn run_with_diagnostics_async<F, Fut, T>(
+    diagnostics_manager: Arc<DiagnosticsManager>,
+    action_kind: ActionKind,
+    f: F,
+) -> Result<T, Status>
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = Result<T, Status>> + Send,
+    T: Send,
+{
+    run_and_track_async(DiagnosticsResultTracker::without_database(&diagnostics_manager, action_kind), f).await
+}
+
+pub async fn run_with_diagnostics_for_database_async<F, Fut, T>(
+    diagnostics_manager: Arc<DiagnosticsManager>,
+    database_name: &str,
+    action_kind: ActionKind,
+    f: F,
+) -> Result<T, Status>
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = Result<T, Status>> + Send,
+    T: Send,
+{
+    run_and_track_async(DiagnosticsResultTracker::with_database(&diagnostics_manager, database_name, action_kind), f)
+        .await
+}
+
+async fn run_and_track_async<F, Fut, T>(mut tracker: DiagnosticsResultTracker<'_>, f: F) -> Result<T, Status>
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = Result<T, Status>> + Send,
+{
+    let result = f().await;
+    if let Err(ref status) = result {
+        tracker.set_error_code(get_status_error_code(status));
+    }
+    result
+}
+
+fn get_status_error_code(status: &Status) -> String {
+    if let Ok(details) = status.check_error_details() {
+        if let Some(error_info) = details.error_info() {
+            return error_info.reason.clone();
+        }
+    }
+    "UKNOWN".to_owned() // TODO: What to do with error codes?
```

Comment:
We should at least use `Status::code()` I think. Although I don't think we should ever be in that situation?

Reply:
Okay, I added replaced it. Wasn't sure if it's correct.